### PR TITLE
Add Cassandra Integration as Vector Database for Phidata

### PR DESCRIPTION
**Pull request by VedantS20**
This PR #1328  introduces a new integration with **Apache Cassandra** as a Vector Database, utilizing **CassIO** for vector storage and retrieval.

#### Dependencies
To use this integration, ensure the following libraries are installed:
```bash
pip install cassio
pip install cassandra-driver
```

#### Cookbook example 
Included in this PR is a cookbook example demonstrating how to initialize and use **CassandraDb** for common vector operations . this is located at  `cookbook/vectordb/cassandraDb.py`

### Additional Notes
Users should configure their Cassandra instance and keyspace before using this integration. [Quick Start Guide](
https://cassandra.apache.org/_/quickstart.html)

### Sample Response
image

Modal : mistral-large
Embedding : mistral-embed

### CODE REVIEW

File: phi/vectordb/cassandra/cassandra.py
Comment (manthanguptaa) [inline phi/vectordb/cassandra/cassandra.py:1]: Your queries and logger.debug statements are formatted in incorrect way which is why they are throwing EOL error when trying to run the code

Comment (VedantS20) [reply phi/vectordb/cassandra/cassandra.py:1]: I think it happened due to my prettier, but I ran the script `./scripts/validate.sh` before submitting the PR . but ig it didn't work..

Fixed it

File: cookbook/vectordb/cassandraDb.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+knowledge_base = PDFUrlKnowledgeBase(
+    urls=["https://phi-public.s3.amazonaws.com/recipes/ThaiRecipes.pdf"],
+    vector_db=CassandraDb(table_name="recipes", keyspace="testkeyspace", session=session, embedder=MistralEmbedder()),
+)
+
+
+knowledge_base.load(recreate=False)  # Comment out after first run
+
+agent = Agent(provider=MistralChat(), knowledge_base=knowledge_base, use_tools=True, show_tool_calls=True)
```
[the remark this answers is not in the document]
Comment (VedantS20) [reply cookbook/vectordb/cassandraDb.py]: Yea , My bad , ig i mistakenly removed it , 

Fixed it..

File: phi/vectordb/cassandra/cassandra.py
Comment (manthanguptaa) [reply phi/vectordb/cassandra/cassandra.py:1]: For formatting we use `./scripts/format.sh`

Comment (VedantS20) [reply phi/vectordb/cassandra/cassandra.py:1]: Yea Sorry, I used `./scripts/format.sh` for formatting . and it work for the rest of the code , but it didn't fixed the logger line . I've fixed it manually

**Comment by manthanguptaa** [comment]
@VedantS20 are you still working on this PR?

**Comment by VedantS20** [comment]
I have resolved all the review comments @manthanguptaa . It is done from my side

**Comment by manthanguptaa** [comment]
@VedantS20 I don't see my comments being addressed. Please do so

**Comment by VedantS20** [comment]
Hey @manthanguptaa , Can you help me recreate this EOL error ? . as I mentioned , format.sh script is not reformatting the logger line . that's why I've fixed it manually . I don't see any validation errors in my files ,
 here is the SS , I'm running both validation and format script.
image

Validation Errors which are there in this image are from mistral.py file, which is not related to this feature . I haven't made any changes in this file .

**Review by manthanguptaa — CHANGES_REQUESTED** [review]

**Comment by VedantS20** [comment]
Thanks @manthanguptaa , I have made the changes . Please Check.

File: phi/vectordb/cassandra/cassandra.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
+from typing import Optional, List, Dict, Any, Iterable
+from phi.vectordb.base import VectorDb
+from phi.embedder import Embedder
+from phi.document import Document
+from .index import PhiMetadataVectorCassandraTable
```
Comment (manthanguptaa) [inline phi/vectordb/cassandra/cassandra.py:5]: Fix the import here. It should look something like this `from phi.vectordb.cassandra.index import PhiMetadataVectorCassandraTable`

File: phi/vectordb/cassandra/index.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+try:
+    from cassio.table.base_table import BaseTable
+    from cassio.table.mixins.metadata import MetadataMixin
+    from cassio.table.mixins.type_normalizer import TypeNormalizerMixin
+    from cassio.table.mixins.vector import VectorMixin
+    from .extra_param_mixin import ExtraParamMixin
+except (ImportError, ModuleNotFoundError):
+    raise ImportError("Could not import cassio python package. " "Please install it with `pip install cassio`.")
```
Comment (manthanguptaa) [inline phi/vectordb/cassandra/index.py:8]: ` raise ImportError("Could not import cassio python package. Please install it with `pip install cassio`.")`

File: cookbook/vectordb/cassandraDb.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
+from phi.agent import Agent
+from phi.knowledge.pdf import PDFUrlKnowledgeBase
+from phi.vectordb.cassandra.cassandra import CassandraDb
+import os
+
+try:
+    from cassandra.cluster import Cluster  # type: ignore
+except (ImportError, ModuleNotFoundError):
+    raise ImportError(
+        "Could not import cassandra-driver python package. " "Please install it with `pip install cassandra-driver`."
```
Comment (manthanguptaa) [inline cookbook/vectordb/cassandraDb.py:10]: `raise ImportError("Could not import cassio python package. Please install it with pip install cassio.")`

File: cookbook/vectordb/cassandraDb.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    )
+from phi.embedder.mistral import MistralEmbedder
+from phi.model.mistral import MistralChat
+
+cluster = Cluster()
+session = cluster.connect("testkeyspace")
```
Comment (manthanguptaa) [inline cookbook/vectordb/cassandraDb.py:16]: Can we move this to the Cassandra class? and take the name as param

**Review by manthanguptaa — CHANGES_REQUESTED** [review]
Few nitpicks. We are close on completing this one

File: cookbook/vectordb/cassandraDb.py (hunk 5 above)
Comment (VedantS20) [reply cookbook/vectordb/cassandraDb.py:16]: Hey @manthanguptaa , The current implementation accepts the session as an argument to the CassandraVectorDb class to allow more flexibility

1- Users can configure and customize their Cassandra session (e.g., authentication, cluster settings, retries) outside the CassandraVectorDb class.
2- If the application already manages Cassandra sessions elsewhere, it avoids redundant session creation and allows sharing a single session across multiple components.

In the cookbook example, I have kept it simple by directly connecting to a local Cassandra cluster with no authentication. However, real-world production scenarios might require more complex connection setups.

Please let me know your thoughts on this approach.

**Review by manthanguptaa — APPROVED** [review]
LGTM! There is one thing we can improve in the future is that the keyspace doesn't get created if it doesn't exist

**Review by ysolanky — APPROVED** [review]

**Pull request merged**